### ISSUE TITLE
isDayBlocked needs a new instance on each render on async operations

### ISSUE THREAD

**Issue by sag1v** [airbnb/react-dates#813]
Hi, consider this scenario:
Rendering `DateRangePicker` normally and passing `isDayBlocked`.
when a user choose checkin-date we want to fetch checkout available dates  from the server and run `isDayBlocked` against the new dates.
In order to achieve this flow, we must provide a new function instance prop of `isDayBlocked` on each render:  
```
<DateRangePicker
    ...
    isDayBlocked={(day) => this.isDayBlocked(day)}
    ...
/> 
```
due to [this logic](https://github.com/airbnb/react-dates/blob/master/src/components/DayPickerRangeController.jsx#L220). 
I'm not sure if this was discussed before but i assume that this behavior can potentially cause a performance hit at some point.    
Is there a way to change the logic that will let us run `isDayBlocked` without passing a new instance and use the same reference of function?
From the top of my head  i can think on one solution, maybe pass additional `bool` that defaults to `false` of course and only when passed as `true` run `isDayBlocked`.  
for example:
```
if (isDayBlocked !== this.props.isDayBlocked || forceBlockCheck) { // keep backwards compatibility 
      this.modifiers['blocked-calendar'] = day => isDayBlocked(day);
      recomputeDayBlocked = true;
 } 
```
I'm not saying this is the best solution, I'm curious to hear what is your take on this issue.

**Comment by ljharb** [comment]
The issue there is that you're passing a new arrow function every render - use a constructor-bound instance method, and you'll get the same `===` function every time.

**Comment by majapw** [comment]
@sag1v @ljharb You can also have multiple `isDayBlocked` methods to force update once a checkin is selected. I imagine something like this would cover these issues.

```
class DateRangePickerWrapper extends React.Component {
  constructor(props) {
    super(props);

    this.isDayBlocked = this.isDayBlocked.bind(this);
    this.isDayBlockedForSelectedCheckin = this.isDayBlockedForSelectedCheckin.bind(this);
  }

  isDayBlocked(day) {
    ...
  }

  isDayBlockedForSelectedCheckin(day) {
    const { checkin } = this.state;
    ...
  }

  render() {
    const { checkin, checkout, focusInput } = this.state;
    const isDayBlocked = (checkin && focusedInput === END_DATE) ? this.isDayBlockedForSelectedCheckin : this.isDayBlocked;

    return (
      <DateRangePicker
        startDate={checkin} 
        endDate={checkout}
        focusedInput={focusedInput}
        isDayBlocked={isDayBlocked}
        ...
       />
    );
  }
}
```

**Comment by sag1v** [comment]
@ljharb I'm actually passing a new instance on purpose so i can force it to run   this block
```
if (isDayBlocked !== this.props.isDayBlocked) {
      this.modifiers['blocked-calendar'] = day => isDayBlocked(day);
      recomputeDayBlocked = true;
 } 
```
This is to bypass the issue where you want to update the blocked days after an aync fetch of the checkout dates. I just don't want to pass a new instance on each render (like I'm doing now) and only pass it when my async operation and callback completed.
I think @majapw has a good point with separating the `isDayBlocked` function in to 2 different functions and pass them respectively to the `focusedInput` state, this way i will only have 2 references and still get my goal.
I need to think about that some more to see if I'm not missing anything about the logic as it should be the same for both checkin and checkout available dates objects.
I think i will try to create a small example (maybe on codeSandbox )  to illustrate this use case and we could talk about it before i close this issue. if you guys don't mind of course! :smiley:

**Comment by sag1v** [comment]
Ok i think @majapw nailed it with this approach. I don't know why i was so fixed on one handler in the first place. 
Anyway [this is a code example](https://codesandbox.io/s/986x52w9o) if you want to see the use case.

**Comment by majapw** [comment]
Glad to hear that you were able to get it working! I think we need some more documentation on the prop modifiers so maybe it would be good to add lessons like these to the README :)

**Comment by sag1v** [comment]
@majapw 

> modifiers so maybe it would be good to add lessons like these to the README :)
Yeah i was thinking the same thing, but in what section? all of the optional `props` doesn't have explanation nor examples.

**Comment by sag1v** [comment]
Sorry but i made a silly mistake, in my code example i kept the syntax like this:
```
const isDayBlocked =
  focusedInput === "startDate"
    ? this.isStartDayBlocked
    : this.isEndDayBlocked;
    //...
isDayBlocked = {(day) => isDayBlocked(day) }
```
So obviously your approach worked as i didn't change the fact that i am passing a new instance of a function on each render.
if i use the desired approach and passing it like this:
```
const isDayBlocked =
  focusedInput === "startDate"
    ? this.isStartDayBlocked
    : this.isEndDayBlocked;
    //...
isDayBlocked = {isDayBlocked}
```
Then the relevant dates are not clickable but they don't display correctly (looks like an available date).
in my code example if you select 02/11/2017 as a start date then the range of 02/11 - 04-11 should be unavailable for an end date.
[link to an example](https://codesandbox.io/s/986x52w9o).
Keep in mind that there isn't an actuall API call only static arrays so i used `setTimeout` to simulate an async call, so only the first date selection should simulate the issue, after that the items are already  stored in the state hence there is no problem.

**Comment by sag1v** [comment]
@majapw @ljharb Have you checked out my example? any thoughts about this behavior? 
Note that i changed the dates collection to December so the check the example you should choose `02/12/2017` instead of `02/11/2017`
